### ISSUE TITLE
Gateway guide?

### ISSUE THREAD

**Issue by enlivenapp** [firesale/firesale#299]
Hi Guys,  Thanks for a pretty awesome product.

I'm building a Pyro site with FireSale (mostly event ticket sales) for a non-profit and they use PlugnPay as the gateway.  Do any of you have a guide for implementing new gateways?

Basically, I'm looking for, where information is coming from within FireSale AND what info is collected. I do have the PlugnPay example code and API info, so I know what is expected from them.  I do realize, that some of this is dependent on which gateway is being used, but any kind of a guide or guidance would be most helpful.

Currently, I'm on Pyro 2.1.5 Pro

Thanks a bunch,
Mike

**Comment by chrisnharvey** [comment]
Hi Mike,

Thanks for your interest in FireSale.

A good place to start would be by looking at the other payment gateways that I have implemented into FireSale and see which gateways have a similar flow to PlugnPay. Finding out whether the card information is taken on the merchant side or on the gateway website will determine how you need to implement the gateway.

I'm more than happy to help you out with this.

**Comment by enlivenapp** [comment]
Hi Chris,

PlugnPay seems to like curl the most, which can get tricky, some servers don't allow it, but allows for a back up where it's a simple cross-site form submission to PlugnPay... so..

cardsave seems to be the closest, PlugnPay only needs a login and password, (no API "key" per se).

The class and default settings method was pretty trivial to figure out, I'm just bumping my head on where the checkout info comes from (through the form view back into the controller to the gateway?), which includes that we could accept the CC info on-site (if using curl) or off-site via PlugnPay's form (which will be easier to implement initially, I think)).

I'm happy to send you the API docs that I have so far, but it won't let me upload here.

**Comment by chrisnharvey** [comment]
Hi Mike,

It may sound strange, but it is in fact easier to implement a gateway that accepts the credit card information on the merchant side (your site), but both ways are relatively easy to implement once you get your head around it.

Feel free to send over the documentation to chris@getfiresale.org and I'll be sure to take a look at them and get back to you with the best approach.

Chris

**Comment by yveveke** [comment]
Controllers / front_cart.php payment function

``` php
// Run payment
                $params = array_merge(array(
                    'notify_url' => site_url($this->routes_m->build_url('cart') . '/callback/' . $gateway . '/' . $order['id']),
                    'return_url' => site_url($this->routes_m->build_url('cart') . '/success'),
                    'cancel_url' => site_url($this->routes_m->build_url('cart') . '/cancel')
                ), $posted_data ? $posted_data : array(), array(
                    'currency_code'  => $this->fs_cart->currency()->cur_code,
                    'amount'         => $this->fs_cart->total() + $order['shipping']['price'],
                    'order_id'       => $this->session->userdata('order_id'),
                    'transaction_id' => $this->session->userdata('order_id'),
                    'reference'      => 'Order #' . $this->session->userdata('order_id'),
                    'description'    => 'Order #' . $this->session->userdata('order_id'),
                    'first_name'     => $order['bill_to']['firstname'],
                    'last_name'      => $order['bill_to']['lastname'],
                    'address1'       => $order['bill_to']['address1'],
                    'address2'       => $order['bill_to']['address2'],
                    'city'           => $order['bill_to']['city'],
                    'region'         => $order['bill_to']['county'],
                    'country'        => $order['bill_to']['country']['code'],
                    'postcode'       => $order['bill_to']['postcode'],
                    'phone'          => $order['bill_to']['phone'],
                    'email'          => $order['bill_to']['email'],
                ));
```

These are the parameters that get posted.
For examples I would suggest to check out code under libraries / gateways.
Don't forget to make a view under views / gateways

**Comment by enlivenapp** [comment]
Chris,  files sent.

yveveke,  Thanks, that definitely helps.

**Comment by enlivenapp** [comment]
I believe I may have this pretty close to worked out.  

I can authorize (only the test card name and number so far) but it connects to the proper site, and provides the correct information and receives a correct response (failed/success/etc)

One thing I'm having a little trouble with is when a payment is successful, there's nothing showing in the Order Status area.  ( ![see image](https://f.cloud.github.com/assets/3036663/266262/232a34e0-8de0-11e2-95e1-53d29d3f9072.png)  I've looked at all of the other gateway files and I'm left slightly perplexed.

Here's the response part of the file: (apparently, I'm too stupid to figure out how to get it to render as code... what's there it's doing all by itself)

class Merchant_plugnpay_response extends Merchant_response
{
    protected $_response_array;

```
public function __construct($response)
{
    $this->_response_array = $response;

    switch ($this->_response_array['FinalStatus'])
    {
        case 'success':
            $this->_status = Merchant_response::COMPLETE;
            $this->_message = lang('merchant_success_response'); // custom, added to language file
            break;

        case 'badcard':
            $this->_status = Merchant_response::FAILED;
            $this->_message = lang('merchant_invalid_card_no');
            break;

        case 'fraud':
            $this->_status = Merchant_response::FAILED;
            $this->_message = lang('merchant_fraud_response');  // custom, added to language file
            break;

        case 'problem':
            $this->_status = Merchant_response::FAILED;
            $this->_message = lang('merchant_payment_failed');          
            break;

            default:
            $this->_status = Merchant_response::FAILED;
            $this->_message = lang('merchant_invalid_response');            
            break;
    }
}
```

}

**Comment by jHoldroyd** [comment]
Just check the order status in the admin, I have a feeling that the missing status on that page is actually a template error that occasionally happens and isn't related to your gateway. If your order status is correct in the backend, you're done.

**Comment by enlivenapp** [comment]
Thanks for the feedback. I was pretty sure a variable wasn't being passed but wanted to double check. 

The admin panel is showing the correct payment status, however, I'm sure customers are going to be confused without a successful payment message. 

I'll start tracking the message and status back through. 

Chris, once I clean up some code, I'm happy to send it on if you're like to add it to the list of available gateways.

**Comment by chrisnharvey** [comment]
Hi Mike,

I understand it would be quite confusing for the customer not being able to see their payment status. A quick fix in your theme should resolve this issue.

I think it'd be a great addition to our list of gateways, so any contribution would be fantastic!

Chris

**Comment by enlivenapp** [comment]
Hey Guys,

Here it is.  Feel free to give me feedback in the issues tab and I'll get changes done.

https://github.com/wbc-mike/PyroCMS-FireSale-Plugnpay-gateway

Mike

**Comment by jHoldroyd** [comment]
It's looking good Mike and if you get it tested a little more and get to a stable build I'm sure Chris would be more than happy to get it integrated. We'll also talk to Adrian (if you want) and see if we can get it pushed up stream to get it integrated into CI merchant and his upcoming composer gateway package so it's there for others to use.

And Chris, by the looks of it those are the default views so it's probably something we need to look into for 1.2.1, I'll open up an issue just to look into it even if that's not the case.

**Comment by enlivenapp** [comment]
I did the open source MIT license on it, so whoever wants to use it is fine, whether it's FireSale or ci-merchant, etc.  it's completely fine with me.  :)

My current server apparently blocks the exec() command from PHP(I get no response at all back), so all testing has failed, although I've run the output through a terminal on the same server and it comes back successful.  Someone may beat me to testing that part, but I do know it's _really_ close, if not correct already.

All I have to do now, is test with a real credit card and have a few other folks lay eyes on it (Chris, etc) just to double check behind me before I'd classify it as 'stable'.

Since I have a specific instance that I'm running and only one way for it to be a completed transaction, I've edited the status and message in the view for complete payments with hardcoded messages (for the time being).  Obviously, I would not recommend this approach, generally.
